seqr/model_utils.py

```diff
@@ -0,0 +1,203 @@
+        "analysis_summary": "analysis_summary_content",
+    },
+    "Individual": {
+        'sex': 'gender',
+        'display_name': 'nickname',
+        'phenotips_eid': 'phenotips_id',
+    },
+    "VariantTagType": {
+        "name": "tag",
+        "description": "title",
+    },
+    "VariantTag": {
+        "variant_tag_type": "project_tag",
+        "created_date": "date_saved",
+        "created_by": "user",
+        "xpos_start": "xpos",
+    },
+    "VariantNote": {
+        "project_tag": "variant_tag_type",
+        "date_saved": "created_date",
+        "user": "created_by",
+        "xpos": "xpos_start",
+    },
+    "LocusList": {
+        "created_by": "owner",
+        "created_date": "last_updated",
+    },
+    "LocusListGene": {
+        "locus_list": "gene_list",
+    },
+}
+
+
+def _update_model(model_obj, **kwargs):
+    for field, value in kwargs.items():
+        setattr(model_obj, field, value)
+
+    model_obj.save()
+
+
+def _is_seqr_model(obj):
+    return type(obj).__module__ == "seqr.models"
+
+
+def find_matching_xbrowse_model(seqr_model):
+    logging.info("find matching xbrowse %s for %s" % (type(seqr_model).__name__, seqr_model))
+    if not _is_seqr_model(seqr_model):
+        raise ValueError("Unexpected model class: %s.%s" % (type(seqr_model).__module__, type(seqr_model).__name__))
+
+    try:
+        seqr_class_name = type(seqr_model).__name__
+
+        if seqr_class_name == "Project":
+            return BaseProject.objects.get(
+                project_id=seqr_model.deprecated_project_id)
+        elif seqr_class_name == "Family":
+            return BaseFamily.objects.get(
+                project__project_id=seqr_model.project.deprecated_project_id,
+                family_id=seqr_model.family_id)
+        elif seqr_class_name == "Individual":
+            return BaseIndividual.objects.get(
+                family__project__project_id=seqr_model.family.project.deprecated_project_id,
+                family__family_id=seqr_model.family.family_id,
+                indiv_id=seqr_model.individual_id)
+
+    except Exception as e:
+        logging.error("ERROR: ", e)
+
+    return None
+
+
+def _convert_seqr_kwargs_to_xbrowse_kwargs(seqr_model, **kwargs):
+    # rename fields
+    seqr_class_name = type(seqr_model).__name__
+    field_mapping = SEQR_TO_XBROWSE_FIELD_MAPPING[seqr_class_name]
+    xbrowse_kwargs = {
+        field_mapping.get(field, field): value for field, value in kwargs.items()
+        if not field_mapping.get(field, field) == _DELETED_FIELD
+    }
+
+    # handle foreign keys
+    for key, value in field_mapping.items():
+        if _is_seqr_model(value):
+            value = find_matching_xbrowse_model(value)
+            if value is not None:
+                field_mapping[key] = value
+            else:
+                logging.info("ERROR: unable to find equivalent seqr model for %s: %s" % (key, value))
+                del field_mapping[key]
+
+    return xbrowse_kwargs
```

Comment:
what do you mean by this?

Reply:
yeah, I think in general we should be trying to rely on explicit foreign key references because its much harder to make mistakes with those rather than having these complex per-model field mappings that are manually defined and any time you change either model you could accidentally break that lookup. Also, for the secondary query I think we should explicitly check that `seqr_locus_list` is null, because if it has a reference to a different model than we don't want it

Reply:
fixed